File: WikiCodeParser/Elements/MdTableElement.cs
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using WikiCodeParser.Nodes;

namespace WikiCodeParser.Elements
{
    public class MdTableElement : Element
    {
        public override bool Matches(Lines lines)
        {
            var value = lines.Value().TrimEnd();
            return value.Length >= 2 && value[0] == '|' && (value[1] == '=' || value[1] == '-');
        }

        public override INode Consume(Parser parser, ParseData data, Lines lines, string scope)
        {
            var arr = new List<TableRow>();
            do
            {
                var value = lines.Value().TrimEnd();
                if (value.Length < 2 || value[0] != '|' || (value[1] != '=' && value[1] != '-')) {
                    lines.Back();
                    break;
                }
                var cells = SplitTable(value.Substring(2)).Select(x => ResolveCell(x, parser, data, scope));
                arr.Add(new TableRow(value[1] == '=' ? "th" : "td", cells));
            } while (lines.Next());

            return new HtmlNode("<table class=\"table table-bordered\">", new NodeCollection(arr), "</table>");
        }

        private static IEnumerable<string> SplitTable(string text)
        {
            var level = 0;
            var last = 0;
            text = text.Trim();
            var len = text.Length;
            int i;
            for (i = 0; i < len; i++)
            {
                var c = text[i];
                if (c == '[') level++;
                else if (c == ']') level--;
                else if ((c == '|' && level == 0) || i == len - 1)
                {
                    yield return text.Substring(last, (i - last) + (i == len - 1 ? 1 : 0));
                    last = i + 1;
                }
            }
            if (last < len) yield return text.Substring(last, (i-last) + (i == len - 1 ? 1 : 0));
        }

        private static INode ResolveCell(string text, Parser parser, ParseData data, string scope)
        {
            var res = Regex.Match(text.Trim(), "^:ref=([a-z ]+)$", RegexOptions.IgnoreCase);
            if (res.Success)
            {
                var name = res.Groups[1].Value;
                var node = data.Get<INode>($"Ref::{name}", () => null);
                if (node != null) return node;
            }
            return parser.ParseTags(data, text, scope, "inline");
        }

        private class TableRow : INode
        {
            public string Type { get; }
            public List<INode> Cells { get; }

            public TableRow(string type, IEnumerable<INode> cells)
            {
                Type = type;
                Cells = cells.ToList();
            }

            public string ToHtml()
            {
                var sb = new StringBuilder("<tr>\n");
                foreach (var cell in Cells)
                {
                    sb.Append("<").Append(Type).Append(">");
                    sb.Append(cell.ToHtml());
                    sb.Append("</").Append(Type).Append(">\n");
                }
                sb.Append("</tr>\n");
                return sb.ToString();
            }

            public string ToPlainText()
            {
                var sb = new StringBuilder();
                var first = true;
                foreach (var cell in Cells)
                {
                    if (!first) sb.Append(" | ");
                    sb.Append(cell.ToPlainText());
                    first = false;
                }
                sb.Append("\n");
                return sb.ToString();
            }

            public IEnumerable<INode> GetChildren()
            {
                return Cells;
            }
        }
    }
}

File: php/src/Elements/MdTableElement.php
<?php

namespace LogicAndTrick\WikiCodeParser\Elements;

use LogicAndTrick\WikiCodeParser\Lines;
use LogicAndTrick\WikiCodeParser\Nodes\HtmlNode;
use LogicAndTrick\WikiCodeParser\Nodes\INode;
use LogicAndTrick\WikiCodeParser\Nodes\NodeCollection;
use LogicAndTrick\WikiCodeParser\Nodes\RefNode;
use LogicAndTrick\WikiCodeParser\ParseData;
use LogicAndTrick\WikiCodeParser\Parser;
use LogicAndTrick\WikiCodeParser\TagParseContext;

class TableRow implements INode
{
    public string $type;
    public array $cells;

    public function __construct(string $type, array $cells)
    {
        $this->type = $type;
        $this->cells = $cells;
    }

    function ToHtml(): string
    {
        $sb = "<tr>\n";
        foreach ($this->cells as $cell) {
            $sb .= "<{$this->type}>{$cell->ToHtml()}</{$this->type}>\n";
        }
        $sb .= "</tr>\n";
        return $sb;
    }

    function ToPlainText(): string
    {
        $sb = '';
        $first = true;
        foreach ($this->cells as $cell) {
            if (!$first) $sb .= ' | ';
            $sb .= $cell->ToPlainText();
            $first = false;
        }
        $sb .= "\n";
        return $sb;
    }

    function GetChildren(): array
    {
        return [...$this->cells];
    }

    function ReplaceChild(int $i, INode $node): void
    {
        $this->cells[$i] = $node;
    }

    function HasContent(): bool
    {
        return true;
    }
}

class MdTableElement extends Element
{
    public function Matches(Lines $lines): bool
    {
        $value = rtrim($lines->Value());
        return strlen($value) >= 2 && $value[0] == '|' && ($value[1] == '=' || $value[1] == '-');
    }

    public function Consume(Parser $parser, ParseData $data, Lines $lines, string $scope): ?INode
    {
        $arr = [];
        do {
            $value = rtrim($lines->Value());
            if (strlen($value) < 2 || $value[0] != '|' || ($value[1] != '=' && $value[1] != '-')) {
                $lines->Back();
                break;
            }
            $cells = array_map(fn (string $x) => self::ResolveCell($x, $parser, $data, $scope), self::SplitTable(substr($value, 2)));
            $arr[] = new TableRow($value[1] == '=' ? 'th' : 'td', $cells);
        } while ($lines->Next());

        return new HtmlNode('<div class="table-responsive"><table class="table table-bordered">', new NodeCollection(...$arr), '</table></div>');
    }

    private static function SplitTable(string $text) : array {
        $ret = [];
        $level = 0;
        $last = 0;
        $text = trim($text);
        $len =  strlen($text);
        for ($i = 0; $i < $len; $i++) {
            $c = $text[$i];
            if ($c == '[') $level++;
            else if ($c == ']') $level--;
            else if (($c == '|' && $level == 0) || $i == $len - 1) {
                $ret[] = trim(substr($text, $last, ($i - $last) + ($i == $len - 1 ? 1 : 0)));
                $last = $i + 1;
            }
        }
        if ($last < $len) $ret[] = trim(substr($text, $last, ($i - $last) + ($i == $len - 1 ? 1 : 0)));
        return $ret;
    }

    private static function ResolveCell(string $text, Parser $parser, ParseData $data, string $scope): INode {
        $success = preg_match('/^:ref=([a-z0-9 ]+)$/i', trim($text), $res);
        if ($success) {
            $name = $res[1];
            return new RefNode($data, $name);
        }
        return $parser->ParseTags($data, $text, $scope, TagParseContext::Block);
    }
}